spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/JdbcPollingChannelAdapter.java

```diff
@@ -119,19 +117,29 @@ public void setSelectSqlParameterSource(SqlParameterSource sqlQueryParameterSour
 	 * The maximum number of rows to pull out of the query results per poll (if
 	 * greater than zero, otherwise all rows will be packed into the outgoing
 	 * message). Default is zero.
-	 *
 	 * @param maxRows the max rows to set
+	 * @deprecated since 5.1 in favor of {@link #setMaxRows(int)}
 	 */
+	@Deprecated
 	public void setMaxRowsPerPoll(int maxRows) {
-		this.maxRowsPerPoll = maxRows;
+		setMaxRows(maxRows);
+	}
+
+	/**
+	 * The maximum number of rows to query. Default is zero - select all records.
+	 * @param maxRows the max rows to set
+	 * @since 5.1
+	 */
+	public void setMaxRows(int maxRows) {
```

Comment:
We should clarify here, and in the docs, that this mechanism is crude - it simply discards the excess results - it is generally better to use the DBMS custom SQL to restrict the number of rows returned, if supported (e.g. `LIMIT n` in MySQL).

Reply:
Yeah... You know I found this in the `JdbcTemplate`:
```
	/**
	 * Set the maximum number of rows for this JdbcTemplate. This is important for
	 * processing subsets of large result sets, avoiding to read and hold the entire
	 * result set in the database or in the JDBC driver if we're never interested in
	 * the entire result in the first place (for example, when performing searches
	 * that might return a large number of matches).
	 * <p>Default is -1, indicating to use the JDBC driver's default configuration
	 * (i.e. to not pass a specific max rows setting on to the driver).
	 * <p>Note: As of 4.3, negative values other than -1 will get passed on to the
	 * driver, in sync with {@link #setFetchSize}'s support for special MySQL values.
	 * @see java.sql.Statement#setMaxRows
	 */
	public void setMaxRows(int maxRows) {
```

From here I'm even starting to think that we don't need our property at all.
Since we have that `Statement` level setting there - see `JdbcTemplate.applyStatementSettings(Statement stmt)`.

Plus of course you are right, that it would be much better to deffer the `MaxRows` to the target vendor `SELECT`.

So, WDYT now ?

Reply:
Do you mean eliminate it altogether and tell them to appropriately configure a `JdbcTemplate` to limit the rows (or use a vendor SQL extension).

Of course, they'd have to use the 4th CTOR (on the ob gateway) for that.

SGTM.

Reply:
Right. Just deprecate the option in favor of such an option on the `JdbcTemplate` to provide.
And also recommend to rely on the  native JDBC vendor limiting support available through SELECT options.

Makes sense ?

Reply:
yes

Reply:
Hm... You know during fixing that I've changed my mind: the current option is per-Statement and doesn't effect `JdbcTempalate` globally. I think we need to keep the current behavior renaming the property. Also I will consider some really per-Statement configuration for optimization.

The limiting with the `SELECT` is still there for documentation...